### PR TITLE
network: Randomly generate a valid subnet for virtual networking. Fix…

### PR DESCRIPTION
**Pull request by townsend2010**
…es #49

### CODE REVIEW

**Comment by townsend2010** [comment]
This branch does a number of things:

1. Checks if QStandardPaths::AppDataLocation/multipassd/vm-ips/multipass_subnet contains a defined subnet.
2. If so, it reuses that and goes on.
3. If not, then it will either use the legacy 10.122.122 subnet if that is already in use or if not, it will generate a random subnet at 10.x.x and validate it.

File: src/platform/backends/qemu/qemu_virtual_machine_factory.cpp [hunk 1]
```diff
@@ -70,135 +71,189 @@ auto generate_tap_device_name(const std::string& vm_name)
...
-    network_cmd.setArguments(args);
+    std::stringstream subnet;
 
-    network_cmd.start();
-    network_cmd.waitForFinished();
+    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::default_random_engine generator(seed);
+    std::uniform_int_distribution<int> distribution(0, 255);
+
+    while (true)
```
Comment (albaguirre) [inline src/platform/backends/qemu/qemu_virtual_machine_factory.cpp]: Should this be finite instead? I'm wary of infinite loops, given the continuing conditions depend on external processes.
Maybe just try 100 possibilities or so?

File: src/platform/backends/qemu/qemu_virtual_machine_factory.cpp [hunk 2]
```diff
@@ -70,135 +71,189 @@ auto generate_tap_device_name(const std::string& vm_name)
...
+    {
+        subnet = std::stringstream();
+        subnet << "10." << distribution(generator) << "." << distribution(generator);
+
+        // Check if well known
+        if (subnet.str() == "10.10.10")
+            continue;
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine_factory.cpp]: I was following what LXD does when it generates a random subnet.  My assumption is that many users use a 10.10.10 local subnet, so it's just categorically skipped.

File: src/platform/backends/qemu/qemu_virtual_machine_factory.cpp (hunk 1 above)
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine_factory.cpp]: Sure, just following what LXD does when it generates a random subnet, but sure, let's bail out after 100 attempts:)

File: src/platform/backends/qemu/qemu_virtual_machine_factory.cpp [hunk 3]
```diff
@@ -70,135 +71,189 @@ auto generate_tap_device_name(const std::string& vm_name)
...
+
+        // Check if used locally
+        if (mp::utils::run_cmd(
+                "bash", {"-c", QString("ip -4 route show | grep -q %1").arg(QString::fromStdString(subnet.str()))}))
+            continue;
+
+        // Attempt to see if used behind the gateway
+        if (mp::utils::run_cmd(
+                "bash",
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine_factory.cpp]: Sure, will fix.

File: src/platform/backends/qemu/qemu_virtual_machine_factory.cpp [hunk 4]
```diff
@@ -70,135 +71,189 @@ auto generate_tap_device_name(const std::string& vm_name)
...
+                {"-c", QString("ping -n -q %1.1 -c 1 -W 1 >/dev/null 2>&1").arg(QString::fromStdString(subnet.str()))}))
+            continue;
+
+        if (mp::utils::run_cmd(
+                "bash",
+                {"-c",
+                 QString("ping -n -q %1.254 -c 1 -W 1 >/dev/null 2>&1").arg(QString::fromStdString(subnet.str()))}))
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine_factory.cpp]: Yep, will change that.

**Comment by townsend2010** [comment]
If we need to include iputils-ping in our snap, should we also include iproute2 and iptables for our calls to 'ip' and 'iproute'?

**Comment by albaguirre** [comment]
@townsend2010 

Yes I think so. I believe any external binary we call should be included as part of the snap.

**Comment by BurlyLuo** [comment]
can we make a self-definition subnet for the network when init the multipass?

**Pull request merged**